x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java

```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
         }
     }
 
+    public static class RoleRestriction implements Writeable, ToXContentObject {
+
+        public static final TransportVersion WORKFLOWS_RESTRICTION_VERSION = TransportVersion.V_8_9_0;
+
+        public static final RoleRestriction NONE = RoleRestriction.builder().build();
+
+        private final String[] workflowsRestriction;
+
+        public RoleRestriction(@Nullable String[] workflowsRestriction) {
+            this.workflowsRestriction = workflowsRestriction == null ? Strings.EMPTY_ARRAY : workflowsRestriction;
+        }
```

Comment:
I like the leniency, but I agree it does not make much sense. I will adjust the parsing and add validation that if  a `restriction` object is provided it must include non-empty `workflows` array. 

This also implies that if a user wants to update an API key and remove workflows restriction it should simply submit update request without `restriction` section. Not saying this is bad. Just wanted to note it down.

e.g.

Create with restriction:
```
"my_restricted_role": {
  "indices": [
    {
      "names": ["books", "movies"],
      "privileges": ["read"]
    }
  ],
  "restriction": {
    "workflows": ["my_workflow"]
  }
}
```
Update to remove restriction:
```
"my_restricted_role": {
  "indices": [
    {
      "names": ["books", "movies"],
      "privileges": ["read"]
    }
  ]
}
```

Reply:
> This also implies that if a user wants to update an API key and remove workflows restriction it should simply submit update request without restriction section. Not saying this is bad. Just wanted to note it down.

I think that's OK. Role descriptors are updated in its entirety. There is no need to explicity say `restriction: {}` or `restriction: null` to remove it. This is just like when you create the API key, you don't need specifiy `restriction: {}` (or `null`) to mean it has no restriction.

 > I'm wondering if there is a convention/reason (from clients perspective) to allow this for workflows as well?

Maybe. But if there is, it is inconsistent. For example, we do not allow null or empty `names` or `privileges` under `indices`. The `field_security` field cannot be null or empty either. If you do prefer it to be a bit more flexible, I can see myself agree with allowing `restriction: {}`. But not `restriction: { workflow: [] }` because it is rather confusing, does it mean no restriction or restricted to be nothing is allowed?